File: bench/cases/alloc/alloc.c
#include <stdio.h>
#include <stdlib.h>

// -DBENCH_ALLOC_STATS builds a counting copy that prints this run's malloc
// count on stderr, so bench/run.sh can compare it against the Bit and Go
// sides of the same case (#3934). The timed binary is built without it and
// is byte-for-byte the program below.
#ifdef BENCH_ALLOC_STATS
static long nmalloc = 0;
#define malloc(n) (nmalloc++, malloc(n))
#define REPORT_ALLOCS() fprintf(stderr, "[allocs] %ld\n", nmalloc)
#else
#define REPORT_ALLOCS() ((void)0)
#endif

typedef struct { long id; } Batch;
typedef struct { long x, y; Batch *owner; } Node;

int main(void) {
  int batches = 2000, per = 5000;
  long total = 0;
  for (int b = 0; b < batches; b++) {
    Batch *owner = malloc(sizeof(Batch));
    owner->id = b;
    Node **nodes = malloc(sizeof(Node *) * per);
    for (int k = 0; k < per; k++) {
      Node *p = malloc(sizeof(Node));
      p->x = b + k;
      p->y = k + 1;
      p->owner = owner;
      nodes[k] = p;
    }
    for (int k = 0; k < per; k++) total += nodes[k]->x + nodes[k]->y + nodes[k]->owner->id;
    for (int k = 0; k < per; k++) free(nodes[k]);
    free(nodes);
    free(owner);
  }
  printf("%ld\n", total);
  REPORT_ALLOCS();
  return 0;
}

File: bench/cases/alloc/alloc.go
package main

import (
	"fmt"
	"os"
	"runtime"
)

type Batch struct{ id int64 }

type Node struct {
	x, y  int64
	owner *Batch
}

func main() {
	batches, per := 2000, 5000
	var total int64 = 0
	for b := 0; b < batches; b++ {
		owner := &Batch{id: int64(b)}
		nodes := make([]*Node, 0, per)
		for k := 0; k < per; k++ {
			nodes = append(nodes, &Node{x: int64(b + k), y: int64(k + 1), owner: owner})
		}
		for _, nd := range nodes {
			total += nd.x + nd.y + nd.owner.id
		}
	}
	fmt.Printf("%d\n", total)
	reportAllocs()
}

// BENCH_ALLOC_STATS=1 prints this run's heap allocation count on stderr, so
// bench/run.sh can compare it against the Bit and C sides of the same case
// (#3934: the three sources drifted into different data structures and the
// row went on comparing them for a day). Off by default and after all timed
// work, so a measured run pays nothing for it.
func reportAllocs() {
	if os.Getenv("BENCH_ALLOC_STATS") == "" {
		return
	}
	var ms runtime.MemStats
	runtime.ReadMemStats(&ms)
	fmt.Fprintf(os.Stderr, "[allocs] %d\n", ms.Mallocs)
}
